backbone/src/main/java/org/researchstack/backbone/ui/task/TaskViewModel.kt

```diff
@@ -209,4 +215,22 @@ internal class TaskViewModel(context: Application, intent: Intent) : AndroidView
     companion object {
         const val TAG = "TaskViewModel"
     }
+
+
+    private fun convertResults(): TaskResult {
```

Comment:
I propose this method to be written like so, what do you think?

```
    private fun updateTaskResultsFrom(clonedResults: TaskResult): TaskResult {
        val stepIds: MutableList<String> = mutableListOf()
        val results = TaskResult(clonedResults.identifier)
        
        task.steps.forEach {
            val result = results.getStepAndResult(it.identifier).second
            if (result != null) {
                results.setStepResultForStep(it, result)
                stepIds.add(it.identifier)
            }
        }

        task.resetCompletedTask(stepIds)
        return results
    }
```

And then change the invocation:

```
            if (isReviewStep(nextStep)) {
                clonedTaskResult?.let {
                    taskResult = updateTaskResultsFrom(it)
                }
                clonedTaskResult = null
...
```

Reply:
^ the "change" (that may be wrong) on my version, is that **if you need to call** `task.resetCompletedTask(list)` regardless of whether there's a "cloned" response or not, then ou should likely change the invocation, so either add an `elvis` operator like:

```
                clonedTaskResult?.let {
                    taskResult = updateTaskResultsFrom(it)
                } ?: task.resetCompletedTask(emptyList())
```

or an `if/null` check, or whatever method you love. But I don't know if calling this is needed (in your version of the code, if clonedTaskResult is null, this would explode.